427 docs rfc add docs about forking services (#4490)
timescale/docs: use-timescale/fork-services.md

---
title: Developer service forks

> excerpt: With developer forks, spin up a branch, run your test, throw it away, or merge it back.

Reviewer: Same comment with respect to the "developer" wording.

Author: Changed to Service forks

products: [cloud]
price_plans: [performance, scale, enterprise]
keywords: [backups, restore]
tags: [recovery, failures]
---

import IntegrationPrereqs from "versionContent/_partials/_integration-prereqs.mdx";
import CLIFORKS from "versionContent/_partials/_devops-cli-service-forks.mdx";

# Developer $SERVICE_SHORT forks

Modern development is highly iterative. Developers and AI agents need safe spaces to test changes before deploying them
to production. Forkable $SERVICE_SHORTs make this natural and easy. Spin up a branch, run your test, throw it away, or

> merge it back.

Reviewer: I guess this explains what "developer" means a bit better, but it wasn't obvious to me at all until I read this, and personally, I'd still think of them as just "Service forks", and the fact that they can be used during development is just one possible use-case. "Developer service forks" makes it sound like a special/different concept, when it's really not (imo).

Author: Changed to Service forks

A fork is an exact copy of a $SERVICE_SHORT at a specific point in time, with its own independent data and configuration,
including:
- The database data and schema
- Configuration
- An admin `tsdbadmin` user with a new password

Forks are fully independent. Changes to the fork don't affect the parent $SERVICE_SHORT, you can query

> them, run migrations, add indexes, or test new features.

Reviewer: The second sentence has a kind of weird flow after the comma. Feels like it should be two separate sentences. Maybe something like this?

```suggestion
Forks are fully independent. Changes to the fork don't affect the parent $SERVICE_SHORT. You can query
them, run migrations, add indexes, or test new features against the fork without affecting the original service.
```

Forks are a powerful way to share production-scale data safely. BI and data science teams often need access to real

> datasets to build models or generate insights. With forkable $SERVICE_SHORTs, you easily create instant, zero-copy

Reviewer: I think we [changed to wording](https://iobeam.slack.com/archives/C07JSQ7SL00/p1760568749145069?thread_ts=1760568728.161159&cid=C07JSQ7SL00) from "instant" to "fast"

```suggestion
datasets to build models or generate insights. With forkable $SERVICE_SHORTs, you easily create fast, zero-copy
```

branches of a production $SERVICE_SHORT that is isolated from production, but contains all the data needed for

> analysis. You share this fork with your analytics teams in seconds. This dramatically reduces friction getting insights

Reviewer: This subject of the sentence changes from plural ("branches") to singular ("is isolated"). Should probably be something more like:

```suggestion
datasets to build models or generate insights. With forkable $SERVICE_SHORTs, you easily create instant, zero-copy
branches of a production $SERVICE_SHORT that are isolated from production, but contain all the data needed for
analysis. You can share these forks with your analytics teams in seconds. This dramatically reduces friction getting insights
```

Author: Done, thanks.

> from live data. 

Reviewer 2: "seconds" is a tad optimistic imo. yes on fluid storage you might get it in 30-40 seconds right now but it's a docs page not an ad so I think we can be a bit more informational in tone.

I also wouldn't focus so much on analytics teams forks are especially useful for testing operations that might require writes (feature development that requires new columns, migrations and new indices) a BI team is often better served by a simple read replica I think.

## Understand $SERVICE_SHORT forks

> You use $SERVICE_SHORT forks for disaster recovery, CI/CD automation, and testing and development. For example, you 

Reviewer: ```suggestion
You can use $SERVICE_SHORT forks for disaster recovery, CI/CD automation, and testing and development. For example, you 
```

can automatically test a major $PG upgrade on a fork before applying it to your production $SERVICE_SHORT. 

$CLOUD_LONG offers the following fork strategies:

- `now`: create a fresh fork of your database at the current time.
   Use when:
   - You need the absolute latest data
   - Recent changes must be included in the fork

- `last-snapshot`: fork from the most recent [automatic backup or snapshot][automatic-backups]. 
  Use when:
  - You want the fastest possible fork creation
  - Slightly behind current data is acceptable

> - `timestamp` - fork from a specific point in time within your [retention period][pricing].

Reviewer 3: ```suggestion
- `timestamp`: fork from a specific point in time within your [retention period][pricing].
```

  Use when:
  - Disaster recovery from a known-good state
  - Investigating issues that occurred at a specific time
  - Testing "what-if" scenarios from historical data

> The retention period for Point-in-time recovery and forking depends on your [pricing plan][pricing-plan-features].

Reviewer 3: ```suggestion
The retention period for point-in-time recovery and forking depends on your [pricing plan][pricing-plan-features].
```

### Fork creation speed

Fork creation speed depends on your pricing plan.

- Free plan: ~30-90 seconds. Uses a Copy-on-Write storage architecture with zero-copy between a fork and the parent.
- Paid plan: varies with the size of your $SERVICE_SHORT, typically 5-20+ minutes. Uses tradional storage architecture

>    with backup restore + WAL replay.

Reviewer: It really depends on the type of _service_ (free or paid), not the pricing plan. Because paid plans are also able to create free services that use the new copy-on-write storage (right??)

Author: Done

### Billing

You can fork a service created in the free plan to a free or a paid plan. However, you cannot fork a service created in 

> a paid plan to a free plan service. 

Reviewer: I'm maybe not the best person to ask, but I feel like this is mixing up the free and paid _plans_ with free vs paid _services_.  I believe the "plan" generally refers to the overall billing plan that the project is on, whereas the important thing here is whether the services _themselves_ are free or paid (because even paid plans can create 2 free services).

- High-performance storage:
  - Copy-on-Write: you are only billed for storage for the chunks that diverge from the parent $SERVICE_SHORT.
  - Traditional: you are billed for storage for the whole $SERVICE_SHORT.
- Object storage tier:

>   - [Tiered data][data-tiering] is shared across forks using Copy-on-Write and traditional storage:

Reviewer: I'm not sure "Copy-on-Write" should be capitalized? Is it a proper noun? Same for other places in this PR.

```suggestion
  - [Tiered data][data-tiering] is shared across forks using copy-on-write and traditional storage:
```

>    - [Tiered data][data-tiering] is shared across forks using Copy-on-Write and traditional storage:

Reviewer 3: Is this colon intentional? If so, the bullets underneath should be more indented?
